_posts/2016-05-30-one-model-to-rule-them-all.md

```diff
@@ -0,0 +1,154 @@
+
+## From Page Models to Neutrons
+
+In order to implement a flexible architecture, we decided to adopt the [Page Object pattern](http://martinfowler.com/bliki/PageObject.html), another widely-used best practice popularized in web test automation. The basic idea behind the page object pattern is to **wrap low-level commands**, usually with a class, exposing methods at the user level. For instance, imagine a login form view with email and password fields plus a submit button. In JavaScript (ES2015), such a class might look like this:
+
+{% highlight javascript %}
+class Login {
+  // Some user-level methods...
+  login(email, password) {
+    return this.setEmail(email).setPassword(password).submit();
+  }
+
+  setEmail(email) {
+    this.driver.element(this.constructor.EMAIL_INPUT).type(email);
+    return this;
+  }
+
+  setPassword(email) { ... }
+  submit() { ... }
+  forgotPassword() { ... }
+}
+
+// ... and some locators to define interface elements.
+Login.EMAIL_INPUT = '#email';
+{% endhighlight %}
+
+Done? Well, not quite. It would be so convenient if elements could be located in either app using the same identifiers, and if the same Appium commands would run flawlessly on both iOS and Android, but this is generally not the case. Today, we don't require this kind of "element ID" synchronization across our two mobile teams, since we want to move as quickly as possible, and introducing this requirement into such a high-velocity development process would only slow everyone down at this phase. We want to **test the application we deploy**, as-is, without requiring special tweaks or interventions from our busy mobile devs.
+
+Fortunately, our model handles these discrepancies by extending the base class:
+
+{% highlight javascript %}
+class Android extends Login { ... }
+
+// No ID on the Android app :(. No problem, XPath for the rescue!
+Android.EMAIL_INPUT = '//android.widget.EditText';
+
+class IOS extends Login {
+  setEmail(email) {
+    // On iOS we can directly inject the value, and it's way faster!
+    this.driver.element(this.constructor.EMAIL_INPUT).setValue(email);
+    return this;
+  }
+}
+{% endhighlight %}
+
+Great! Now we can take into account platform specificities while having most of our code shared. Now, what if we want to reuse our login form model for a signup form? We had to rethink the Page Object pattern and come up with a new paradigm. Using a mix-in-like construction, we can inject new methods from the `SignUp` class into parent classes to create new classes which benefit from previously-implemented behaviors and locators. `SignUp` works as an interface with default methods:
```

Comment:
You can say "mixin-like" instead of "mix-in-like" ("mixin" isn't hyphenated, normally).